scripts/build-tools.sh

```diff
@@ -5,69 +5,123 @@
 # fail immediately on any errors
 set -e
 
-
 print_usage()
 {
         cat <<EOFUSAGE
-usage: $0 [-t|-f]
+usage: $0 [-c|-f|-h|-l|-p|-t|-T]
+       [-c] Build ctl
+       [-f] Build fuzzer
+       [-h] Display help
+       [-l] Build logger
+       [-p] Build probes
        [-t] Build test topologies
-       [-f] Build fuzzer"
+       [-T] Build topologies
+       [-n] No build, only cmake
 EOFUSAGE
 }
 
-build_tools()
+# generate Makefiles
+prepare_build()
 {
-        cd "$SOF_REPO/tools"
-        rm -rf build_tools
-        mkdir build_tools
-        cd build_tools
+        rm -rf "$BUILD_TOOLS_DIR"
+        mkdir "$BUILD_TOOLS_DIR"
+
+        cd "$BUILD_TOOLS_DIR"
         cmake ..
-        make -j "$NO_PROCESSORS"
+
+        mkdir "$BUILD_TOOLS_DIR/fuzzer"
+        cd "$BUILD_TOOLS_DIR/fuzzer"
+        cmake ../../fuzzer
 }
 
-build_test()
+make_tool()
 {
-        cd "$SOF_REPO/tools/build_tools"
-        make tests -j "$NO_PROCESSORS"
+        # if no argument provided, all the tools will be built. Empty tool is
+        # okay.
+        tool=$1
+        make -C "$BUILD_TOOLS_DIR" -j "$NO_PROCESSORS" $tool
 }
 
-build_fuzzer()
+make_fuzzer()
 {
-        cd "$SOF_REPO/tools/build_tools"
-        mkdir build_fuzzer
-        cd build_fuzzer
-        cmake ../../fuzzer
-        make -j "$NO_PROCESSORS"
+        make -C "$BUILD_TOOLS_DIR/fuzzer" -j "$NO_PROCESSORS"
+}
+
+print_build_info()
+{
+       cat <<EOFUSAGE
+
+Build commands for respective tools:
+        ctl:        make -C "$BUILD_TOOLS_DIR" sof-ctl
+        logger:     make -C "$BUILD_TOOLS_DIR" sof-logger
+        probes:     make -C "$BUILD_TOOLS_DIR" sof-probes
+        tests:      make -C "$BUILD_TOOLS_DIR" tests
+        topologies: make -C "$BUILD_TOOLS_DIR" topologies
+        fuzzer:     make -C "$BUILD_TOOLS_DIR/fuzzer"
+EOFUSAGE
 }
 
 main()
 {
-        local DO_BUILD_TEST DO_BUILD_FUZZER SCRIPT_DIR SOF_REPO
+        local DO_BUILD_ctl DO_BUILD_fuzzer DO_BUILD_logger DO_BUILD_probes \
+                DO_BUILD_test DO_BUILD_topology SCRIPT_DIR SOF_REPO DO_BUILD
         SCRIPT_DIR=$(cd "$(dirname "$0")" && pwd)
         SOF_REPO=$(dirname "$SCRIPT_DIR")
+        BUILD_TOOLS_DIR="$SOF_REPO"/tools/build_tools
         : ${NO_PROCESSORS:=$(nproc)}
 
-        DO_BUILD_TEST=false
-        DO_BUILD_FUZZER=false
-        while getopts "tf" OPTION; do
+        prepare_build
+
+        if [ $# -eq 0 ]; then
+                if "$DO_BUILD"; then
+                        make_tool
+                        make_fuzzer
+                        exit
+                fi
```

Comment:
It's very rare that adding a negation helps but in this particular case I find this easier to understand:
```shell
if ! "$CMAKE_ONLY"; then
```

Reply:
actually removing the conditional altogether. first, `DO_BUILD` is undefined until L89 so `./scripts/build-tools.sh` fails. And when we do specify `-n` the number of options will never be 0, hence, this will never execute, so the additional conditional is unnecessary.

Reply:
Thx I missed one `exit`. Funny how easy it is to get lost in not that much cyclomatic complexity. I think your latest version 1c3dfde303489f reduces it.